pkg/controller/webhook/webhook_controller.go

```diff
@@ -0,0 +1,170 @@
+package webhook
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/golang/glog"
+
+	admission_v1beta1 "k8s.io/api/admission/v1beta1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+
+	shipper "github.com/bookingcom/shipper/pkg/apis/shipper/v1alpha1"
+)
+
+type Controller struct {
+	bindAddr string
+	bindPort string
+
+	tlsCertFile       string
+	tlsPrivateKeyFile string
+}
+
+var (
+	runtimeScheme = runtime.NewScheme()
+	codecs        = serializer.NewCodecFactory(runtimeScheme)
+	deserializer  = codecs.UniversalDeserializer()
+)
+
+func NewController(bindAddr, bindPort, tlsPrivateKeyFile, tlsCertFile string) *Controller {
+	return &Controller{
+		bindAddr:          bindAddr,
+		bindPort:          bindPort,
+		tlsPrivateKeyFile: tlsPrivateKeyFile,
+		tlsCertFile:       tlsCertFile,
+	}
+}
+
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	addr := c.bindAddr + ":" + c.bindPort
+	mux := c.initializeHandlers()
+	server := &http.Server{
+		Addr:    addr,
+		Handler: mux,
+	}
+
+	go func() {
+		var serverError error
+		if c.tlsCertFile == "" || c.tlsPrivateKeyFile == "" {
+			serverError = server.ListenAndServe()
+		} else {
+			serverError = server.ListenAndServeTLS(c.tlsCertFile, c.tlsPrivateKeyFile)
+		}
+
+		if serverError != nil {
+			glog.Fatalf("failed to start shipper-webhook-controller: %v", serverError)
+		}
+	}()
+
+	glog.V(2).Info("Started WebHook controller")
+
+	<-stopCh
+
+	glog.V(2).Info("Shutting down WebHook controller")
+
+	if err := server.Shutdown(context.Background()); err != nil {
```

Comment:
This will trigger the `Fatalf` on line 60.

Reply:
Fixed.